pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -253,26 +307,95 @@ type dynamicResources struct {
 	enabled                    bool
 	fh                         framework.Handle
 	clientset                  kubernetes.Interface
-	claimLister                resourcev1alpha2listers.ResourceClaimLister
-	classLister                resourcev1alpha2listers.ResourceClassLister
+	listers                    listers
+	indexers                   indexers
 	podSchedulingContextLister resourcev1alpha2listers.PodSchedulingContextLister
+
+	// claimAssumeCache enables temporarily storing a newer claim object
+	// while the scheduler has allocated it and the corresponding object
+	// update from the apiserver has not been processed by the claim
+	// informer callbacks. Claims get added here in Reserve and removed by
+	// the informer callback (based on the "newer than" comparison in the
+	// assume cache) or when the API call in PreBind fails.
+	//
+	// It uses cache.MetaNamespaceKeyFunc to generate object names, which
+	// therefore are "<namespace>/<name>".
+	//
+	// This is necessary to ensure that reconstructing the resource usage
+	// at the start of a pod scheduling cycle doesn't reuse the resources
+	// assigned to such a claim. Alternatively, claim allocation state
+	// could also get tracked across pod scheduling cycles, but that
+	// - adds complexity (need to carefully sync state with informer events
+	//   for claims and NodeResourceSlices)
+	// - would make integration with cluster autoscaler harder because it would need
+	//   to trigger informer callbacks.
+	//
+	// When implementing cluster autoscaler support, this assume cache or
+	// something like it (see https://github.com/kubernetes/kubernetes/pull/112202)
+	// might have to be managed by the cluster autoscaler.
+	claimAssumeCache volumebinding.AssumeCache
```

Comment:
Implementing CA support in 1.31 makes sense to me, there's no way we implement it by the 1.30 code freeze.

I think it should work, roughly like this:

CA's cluster snapshot implements a subset of `AssumeCache` needed by this plugin:
  * The snapshot keeps a real claim lister, plus a stack of in-memory collections of assumed objects (the stack is needed for forking).
  * AssumeCache.Assume(obj): add the object to the in-memory collection on top of the stack
  * AssumeCache.Restore(objName): remove the object from the in-memory collection on top of the stack
  * AssumeCache.List(): call the lister, add/update the objects from all in-memory collection from the stack to the list (objects from collections closer to the top of the stack overwrite objects from lower collections), return
  * ClusterSnapshot.Fork(): Add another in-memory collection to the stack
  * ClusterSnapshot.Revert(): Pop the top in-memory collection from the stack

Essentially in CA each simulation would create a new in-memory collection for Assumed claims while Forking the cluster snapshot, and then drop it after finishing the simulation (plus the stack part for handling nested simulations). The plugin would never encounter the assumed updates in a real lister.

The plugin would have to always go through the cache while accessing Claims for the purpose of deciding whether a pod fits on a node, and not expect the Assumes to be eventually reflected in the real lister. I'm not sure if that's true in the current version, given that PreFilter uses the real lister?

Of course we'd need the cache to be injectable into the framework/plugin somehow.

WDYT? I'll also have to validate this with other CA maintainers, early next week.

Reply:
This sounds right to me. You will also need to figure out how to call Reserve and how to simulate NodeResourceSlices. That will cover scale up.

For scale down, you will have to built support into CA for clearing `ReservedFor` and `Allocation` as pods get evicted, similar to how it is done in `pkg/controller/resourceclaim/controller.go`.

> The plugin would have to always go through the cache while accessing Claims for the purpose of deciding whether a pod fits on a node, and not expect the Assumes to be eventually reflected in the real lister.  I'm not sure if that's true in the current version, given that PreFilter uses the real lister?

This needs to be changed. The problem right now is that the event mechanism is triggered by changes in the informer cache. What we want instead is that, for claims, it gets triggered by changes in the assume cache. But for that we need to extend the assume cache so that it can register and invoke callbacks, similar to the informer cache. Then the flow of changes will be `informer -> assume cache -> event handlers`. Then when an event handler lists claims in the cache, it can be sure that the cache has already been updated. That is not the case right now.

There's another TODO about this in `isSchedulableAfterClaimChange` because there's a very small race condition around starting and stopping pods so quickly that the informer cache hasn't had a chance to catch up. I doubt that this can really happen, therefore I think it's okay to work on this in the 1.31 time frame together with the CA implementation. I can take this part.

Reply:
Yup, the idea is to call `Reserve()` any time CA "schedules" a Pod onto a Node in the cluster snapshot. Looks like we can just use `RunReservePluginsReserve` from the framework here.

As far as clearing `ReservedFor` and `Allocation` when removing a pod from a node, am I getting it right that for CA it boils down to:
* Always remove the pod from `ReservedFor`.
* If the pod was the only entry in `ReservedFor`, clear `Allocation`.

As long as the logic is similarly simple to the above (and not likely to change), we can just replicate it in CA and we should be good.

I'm not sure if the informer callbacks are a problem for CA. IIUC, they only affect whether scheduler should evaluate the pod again. CA has a completely different pod processing algorithm, and doesn't take the regular scheduler queueing logic into account at all while doing its simulations.

From CA's standpoint, all that matters is that the logic in `PreFilter`, `Filter`, and `Reserve` only operates on the cache for the purpose of scheduling decisions (i.e. `PreFilter` and `Filter` use the cache state for answering, `Reserve` reflects the decision in the cache). This is definitely true for `Reserve`, and seems mostly true for `PreFilter` and `Filter`, but it's definitely not obvious (the real lister is used a lot, but it seems that the scheduling decisions themselves are based on the cache?).

As far as simulating `NodeResourceSlices` goes, I think we ultimately also want to simulate `ResourceClaimParameters` (and any other pod-scoped DRA object). CA (or at least some downstream implementations) sometimes inject fake pods into the simulation, based on some other pods. For this use-case, we'd need to similarly inject fake `ResourceClaimParameters`.

Technically, AFAIU we'd need to plumb all such simulated objects through `framework.SharedLister`, and the plugin would have to use them instead of the ones from `SharedInformerFactory`. For the first part, it seems easiest to just add `NodeResourceSlices` to `NodeInfo`, and `ResourceClaimParameters` to `PodInfo`. Not sure if there are any potential problems with this on the plugin side.

@alculquicondor Could you validate the plan for 1.31 scheduler-CA integration in this thread?

Reply:
Without looking into the details, it sounds like we have a reasonable path forward for implementing CA support, even if it requires some minor changes to the scheduler code.

So I agree with the plan.